File: windows/ShareWindows/Protocol/Messages.cs
using System;
using System.Buffers.Binary;
using System.Collections.Generic;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace MeshDrop.Protocol;

public sealed record HelloMessage(
    [property: JsonPropertyName("id")] string Id,
    [property: JsonPropertyName("name")] string Name,
    [property: JsonPropertyName("os")] string Os,
    [property: JsonPropertyName("model")] string? Model,
    [property: JsonPropertyName("fp")] string Fp,
    [property: JsonPropertyName("protocol_versions")] List<byte> ProtocolVersions);

public sealed record HelloAckMessage(
    [property: JsonPropertyName("id")] string Id,
    [property: JsonPropertyName("name")] string Name,
    [property: JsonPropertyName("os")] string Os,
    [property: JsonPropertyName("model")] string? Model,
    [property: JsonPropertyName("fp")] string Fp,
    [property: JsonPropertyName("protocol_versions")] List<byte> ProtocolVersions,
    [property: JsonPropertyName("selected_version")] byte SelectedVersion);

public sealed record TextMessage(
    [property: JsonPropertyName("id")] string Id,
    [property: JsonPropertyName("content")] string Content,
    [property: JsonPropertyName("ts")] long Ts);

public sealed record FileMeta(
    [property: JsonPropertyName("index")] int Index,
    [property: JsonPropertyName("name")] string Name,
    [property: JsonPropertyName("size")] long Size,
    [property: JsonPropertyName("sha256")] string Sha256);

public sealed record FileOfferMessage(
    [property: JsonPropertyName("transfer_id")] string TransferId,
    [property: JsonPropertyName("files")] List<FileMeta> Files);

public sealed record FileAcceptMessage(
    [property: JsonPropertyName("transfer_id")] string TransferId,
    [property: JsonPropertyName("index")] int Index,
    [property: JsonPropertyName("resume_offset")] long ResumeOffset);

public sealed record FileRejectMessage(
    [property: JsonPropertyName("transfer_id")] string TransferId,
    [property: JsonPropertyName("index")] int Index,
    [property: JsonPropertyName("reason")] string Reason);

public sealed record FileCompleteMessage(
    [property: JsonPropertyName("transfer_id")] string TransferId,
    [property: JsonPropertyName("index")] int Index);

public static class MessageCodec
{
    private static readonly JsonSerializerOptions s_options = new()
    {
        PropertyNamingPolicy = null,    // 已用 JsonPropertyName 显式声明
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
    };

    public static byte[] Encode<T>(T value) =>
        Encoding.UTF8.GetBytes(JsonSerializer.Serialize(value, s_options));

    public static T Decode<T>(byte[] body) =>
        JsonSerializer.Deserialize<T>(Encoding.UTF8.GetString(body), s_options)
        ?? throw new InvalidOperationException("decode null");
}

/// <summary>
/// FILE_CHUNK 二进制头部：
/// [transfer_id (16)][index (u32 BE)][offset (u64 BE)][data...]
/// </summary>
public readonly record struct FileChunkHeader(Guid TransferId, uint Index, ulong Offset)
{
    public const int Size = 16 + 4 + 8;

    public static byte[] Encode(FileChunkHeader header, ReadOnlySpan<byte> data)
    {
        var buf = new byte[Size + data.Length];
        // .NET Guid 的字节顺序：前 4 字节是 Data1 (LE int32)，要转为我们规范的 big-endian
        // 用 ToByteArray(bigEndian: true) 保证全 BE
        var guidBytes = header.TransferId.ToByteArray(bigEndian: true);
        guidBytes.CopyTo(buf, 0);
        BinaryPrimitives.WriteUInt32BigEndian(buf.AsSpan(16, 4), header.Index);
        BinaryPrimitives.WriteUInt64BigEndian(buf.AsSpan(20, 8), header.Offset);
        data.CopyTo(buf.AsSpan(Size));
        return buf;
    }

    public static (FileChunkHeader header, byte[] data)? Decode(ReadOnlySpan<byte> body)
    {
        if (body.Length < Size) return null;
        var transferId = new Guid(body.Slice(0, 16), bigEndian: true);
        var index = BinaryPrimitives.ReadUInt32BigEndian(body.Slice(16, 4));
        var offset = BinaryPrimitives.ReadUInt64BigEndian(body.Slice(20, 8));
        var data = body.Slice(Size).ToArray();
        return (new FileChunkHeader(transferId, index, offset), data);
    }
}

File: android/app/src/main/java/com/welape/meshdrop/protocol/Messages.kt
package com.welape.meshdrop.protocol

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.UUID

@Serializable
data class HelloMessage(
    val id: String,
    val name: String,
    val os: String,
    val model: String?,
    val fp: String,
    val protocol_versions: List<Int>,
)

@Serializable
data class HelloAckMessage(
    val id: String,
    val name: String,
    val os: String,
    val model: String?,
    val fp: String,
    val protocol_versions: List<Int>,
    val selected_version: Int,
)

@Serializable
data class TextMessage(
    val id: String,
    val content: String,
    val ts: Long,
)

@Serializable
data class FileMeta(
    val index: Int,
    val name: String,
    val size: Long,
    val sha256: String,
)

@Serializable
data class FileOfferMessage(
    val transfer_id: String,
    val files: List<FileMeta>,
)

@Serializable
data class FileAcceptMessage(
    val transfer_id: String,
    val index: Int,
    val resume_offset: Long,
)

@Serializable
data class FileRejectMessage(
    val transfer_id: String,
    val index: Int,
    val reason: String,
)

@Serializable
data class FileCompleteMessage(
    val transfer_id: String,
    val index: Int,
)

@Serializable
data class FileCancelMessage(
    val transfer_id: String,
    val index: Int? = null,
    val reason: String,
)

object MessageCodec {
    val json: Json = Json {
        ignoreUnknownKeys = true
        encodeDefaults = false
    }

    inline fun <reified T> encode(value: T): ByteArray =
        json.encodeToString(kotlinx.serialization.serializer<T>(), value).toByteArray(Charsets.UTF_8)

    inline fun <reified T> decode(body: ByteArray): T =
        json.decodeFromString(kotlinx.serialization.serializer<T>(), body.toString(Charsets.UTF_8))
}

/**
 * FILE_CHUNK 的二进制头部 + 数据。规范见 protocol/messages.md。
 *
 * ```
 * | transfer_id (16) | index (u32 BE) | offset (u64 BE) | data ... |
 * ```
 */
data class FileChunkHeader(
    val transferId: UUID,
    val index: Int,
    val offset: Long,
) {
    companion object {
        const val SIZE = 16 + 4 + 8

        fun encode(header: FileChunkHeader, data: ByteArray): ByteArray {
            val buf = ByteBuffer.allocate(SIZE + data.size).order(ByteOrder.BIG_ENDIAN)
            buf.putLong(header.transferId.mostSignificantBits)
            buf.putLong(header.transferId.leastSignificantBits)
            buf.putInt(header.index)
            buf.putLong(header.offset)
            buf.put(data)
            return buf.array()
        }

        fun decode(body: ByteArray): Pair<FileChunkHeader, ByteArray>? {
            if (body.size < SIZE) return null
            val buf = ByteBuffer.wrap(body).order(ByteOrder.BIG_ENDIAN)
            val msb = buf.long
            val lsb = buf.long
            val index = buf.int
            val offset = buf.long
            val data = body.copyOfRange(SIZE, body.size)
            return Pair(FileChunkHeader(UUID(msb, lsb), index, offset), data)
        }
    }
}
